### PR TITLE
[IMP] Spreadsheet: Introduce single-value INDEX function

### PR DESCRIPTION
**Pull request by dhrp-odoo**
## Description:

This PR introduces a version of the INDEX function. The new INDEX function is designed to return single values, addressing a specific use case where array output is not required.

Task: : [3483842](https://www.odoo.com/web#id=3483842&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_lt("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: src/functions/module_lookup.ts [hunk 1]
```diff
@@ -127,6 +130,57 @@ export const HLOOKUP: AddFunctionDescription = {
...
 };
 
+// -----------------------------------------------------------------------------
+// INDEX
+// -----------------------------------------------------------------------------
+
+export const INDEX: AddFunctionDescription = {
+  description: _lt(`Returns the content of a cell, specified by row and column offset.`),
+  args: args(`
+      reference (range) ${_lt("The range of cells from which the value is returned.")}
```
Comment (hokolomopo) [inline src/functions/module_lookup.ts:140]: A reason you changed from (any, range) to (range) ?

File: tests/functions/module_lookup.test.ts [hunk 2]
```diff
@@ -954,3 +954,57 @@ describe("XLOOKUP formula", () => {
...
+    expect(grid.B5).toBe(grid.B3);
+    expect(grid.C5).toBe(grid.C2);
+    expect(grid.D5).toBe(grid.A3);
+    expect(grid.E5).toBe("#ERROR");
+    expect(grid.F5).toBe("#ERROR");
+    expect(grid.G5).toBe("#ERROR");
+  });
+
+  test("take format into account", () => {
+    const grid = {
```
Comment (hokolomopo) [inline tests/functions/module_lookup.test.ts:1001]: keep original formatting with prettier-ignore

File: tests/functions/module_lookup.test.ts [hunk 3]
```diff
@@ -954,3 +954,51 @@ describe("XLOOKUP formula", () => {
...
+      A1: "3",    B1: "Hello",  C1: "=SUM(A1:A3)",
+      A2: "333",  B2: "World",  C2: "=GT(A3,A1)",
+      A3: "33",   B3: "!",      C3: "=CONCAT(B3,B2)",
+      A5: "=INDEX(A1:A3, 1, 1)",
+      B5: "=INDEX(B1:B3, 3, 1)",
+      C5: "=INDEX(C1:C3, 2, 1)",
+      D5: "=INDEX(A1:A3, 3, 1)",
+      E5: "=INDEX(A1:A3, 0, 0)",
+      F5: "=INDEX(A1:A3, 4, 1)",
+      G5: "=INDEX(A1:A3, 1, 4)",
```
Comment (hokolomopo) [inline tests/functions/module_lookup.test.ts:983]: The whole point of this formatting with //prettier-ignore is to have the structure look like a real grid. It kinda deafeat the purpose if you add cells that aren't well aligned.

But that's useful for visualising the argument "range" (A1:C3), but we don't really care for the rest. Change the tests A5, B5, C5... to A5, A6, A7... and we should be good

File: src/functions/module_lookup.ts [hunk 4]
```diff
@@ -127,6 +130,57 @@ export const HLOOKUP: AddFunctionDescription = {
...
+      )}
+`),
+  returns: ["ANY"],
+  computeFormat: (reference: Arg, row: PrimitiveArg, column: PrimitiveArg) => {
+    const _row = toNumber(row.value);
+    const _column = toNumber(column.value);
+    return reference[_column - 1][_row - 1]?.format;
+  },
+  compute: function (
+    reference: MatrixArgValue,
```
Comment (hokolomopo) [inline src/functions/module_lookup.ts:155]:
```suggestion
    reference: ArgValue,
```

File: tests/functions/module_lookup.test.ts (hunk 3 above)
Comment (dhrp-odoo) [reply tests/functions/module_lookup.test.ts]: Ohh, Okay!!
I've made the suggested changes.

File: src/functions/module_lookup.ts [hunk 5]
```diff
@@ -127,6 +131,53 @@ export const HLOOKUP: AddFunctionDescription = {
...
+      )}
+`),
+  returns: ["ANY"],
+  computeFormat: (reference: Arg, row: PrimitiveArg, column: PrimitiveArg) => {
+    const _row = toNumber(row.value);
+    const _column = toNumber(column.value);
+    return reference[_column - 1][_row - 1]?.format;
+  },
+  compute: function (reference: ArgValue, row: PrimitiveArgValue, column: PrimitiveArgValue): any {
+    const _reference = (isMatrix(reference) ? reference : [[reference]]) as ArgValue[][];
```
Comment (hokolomopo) [inline src/functions/module_lookup.ts:156]: If you copied the typing of `isMatrix` correctly you wouldn't have to cast the result (your cast is wrong btw).

```suggestion
    const _reference = isMatrix(reference) ? reference : [[reference]]);
    
export function isMatrix(x: any): x is any[][] {
  return Array.isArray(x) && Array.isArray(x[0]);
}

```

**Review by hokolomopo — COMMENTED** [review]
Last fix then I can r+

**Review by hokolomopo — COMMENTED** [review]
@robodoo r+

Thanks for the work 👍

**Pull request closed without merging**